Bluetooth: Mesh: Add comp p129

Andrewpini opened the pull request:
- Adds support for composition data page 129.
- Adds testing of composition data page 129 in the
mesh_prov_pst_pb_remote_ncrp bsim test.

alxelax commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2086,15 +2096,30 @@ int bt_mesh_comp_store(void)
+	}
+
+	net_buf_simple_reset(&buf);
+	err = bt_mesh_comp_data_get_page_1(&buf);
```
Shouldn't this be present only if remote provisioning feature is present?

Andrewpini replied:
Im not sure if the RPR server must be present, but I think `CONFIG_BT_MESH_COMP_PAGE_1` clause should be added.

alxelax commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1892,7 +1893,16 @@ static int comp_set(const char *name, size_t len_rd, settings_read_cb read_cb,
 	return 0;
 }
-BT_MESH_SETTINGS_DEFINE(comp, "cmp", comp_set);
+BT_MESH_SETTINGS_DEFINE(comp, "cmp0", comp_p0_set);
```
Seems this breaks backward compatibility. Could you apply the same trick that Pavel did for virtual address support?
https://github.com/zephyrproject-rtos/zephyr/pull/57878/files#diff-007d797c63a5cbfd92a1922fc7fe3e40502988f252054ebeb6da1d77e5f25e87R334
https://github.com/zephyrproject-rtos/zephyr/pull/57878/files#diff-fced093b1243729a3f60760434e674f6c075f19c42b1e71aa44b8c1eb63721eeR1742

Andrewpini replied:
But if I reinstate the `"cmp"` name this is not an issue, right? I did not think of the backwards compabillity when I did this, but if it causes issues changing the name I can avoid changing the name.

alxelax replied:
In general, I like your change. Page index is a good idea. The only issue is backward compatibility.

PavelVPV replied:
FYI. There is no issue with backward compatibility. This code is part of mesh-1.1 features. Mesh-1.1 is still experimental. No one should have been using it in production yet. So it is absolutely fine to change the way CDP0 is stored.

alxelax commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2140,16 +2166,23 @@ static int read_comp_cb(const char *key, size_t len, settings_read_cb read_cb,
 	}
 
-	err = settings_load_subtree_direct("bt/mesh/cmp", read_comp_cb, buf);
+	sprintf(subtree, "bt/mesh/cmp%d", page);
```
Do you really need `sprintf` and `stdio` for this? How about concatenation?
`printf` family of functions is quite heavy.

Andrewpini replied:
I will look into this 👍

alxelax commented on subsys/bluetooth/mesh/shell/cfg.c:
```diff
@@ -153,7 +154,7 @@ static int cmd_get_comp(const struct shell *sh, size_t argc, char *argv[])
 		}
 	}
 
-	if (IS_ENABLED(CONFIG_BT_MESH_COMP_PAGE_1) && page == 1) {
```
Why did you remove conditional compilation?

Andrewpini replied:
I will reinstate it 👍

Andrewpini replied:
Oh, right. This is implementation on cfg CLI side. As far as I can see `CONFIG_BT_MESH_COMP_PAGE_1` option only applies to/effects the cfg SRV side.

alxelax commented on tests/bsim/bluetooth/mesh/prj_mesh1d1.conf:
```diff
@@ -61,6 +61,8 @@ CONFIG_BT_MESH_DFD_SRV=y
 CONFIG_BT_MESH_PRIV_BEACON_SRV=y
 CONFIG_BT_MESH_PRIV_BEACON_CLI=y
+CONFIG_BT_MESH_COMP_PAGE_1=y
+CONFIG_BT_MESH_MODEL_EXTENSIONS=y
```
It has already been declared here. Line 45

alxelax reviewed (commented):
Need more work. I will not block it since I'm on vacation next 3 weeks. Hope you consider my comments.

Andrewpini committed:
Bluetooth: Mesh: Support for comp data page 129

Adds support for composition data page 129.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
tests: Bluetooth: Mesh: Add comp p129 test to bsim

Adds testing of composition data page 129 in the
mesh_prov_pst_pb_remote_ncrp bsim test.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

alxelax commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -7,6 +7,7 @@
 #include <zephyr/kernel.h>
 #include <errno.h>
 #include <stdlib.h>
+#include <stdio.h>
```
do not need this header anymore

Andrewpini replied:
My bad, will remove

alxelax commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2114,10 +2139,11 @@ static void comp_data_clear(void)
 	int err;
 
 	err = settings_delete("bt/mesh/cmp");
+	err |= settings_delete("bt/mesh/cmp1");
```
It is better to keep them separate. err1 and err2 maybe? When you see some strange value as LOG output it difficult differentiate what was initial error in case of both.

alxelax commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -75,27 +79,66 @@ static int dev_comp_data_get(struct bt_mesh_model *model,
+		}
+		break;
+
+	case 1:
```
Again shouldn't these cases (1, 129) be wrapped around:
```
#if IS_ENABLED(CONFIG_BT_MESH_COMP_PAGE_1)
case 1:
.....
break;
case 129:
....
break;
#endif
```

alxelax commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -75,27 +79,66 @@ static int dev_comp_data_get(struct bt_mesh_model *model,
+		break;
+
+	default:
+		/*Should never happen*/
```
Why should it never happen? What if a client requested a page that the server doesn't support? What server should reply?

Andrewpini replied:
Discussed in person

PavelVPV commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -75,27 +79,69 @@ static int dev_comp_data_get(struct bt_mesh_model *model,
+#endif
+
+	default:
+		/*Should never happen*/
```
```suggestion
		/* Should never happen */
```

PavelVPV commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -75,27 +79,69 @@ static int dev_comp_data_get(struct bt_mesh_model *model,
-	if (IS_ENABLED(CONFIG_BT_MESH_COMP_PAGE_1) && page == 1) {
-		err = bt_mesh_comp_data_get_page_1(&sdu);
+
+	switch (page) {
```
The previous `if` looked a bit more compact. Or it could be replaced with something like this:
```
static int comp_data_get_page(struct net_buf_simple *buf, size_t page, size_t offset)
{
	if (atomic_test_bit(bt_mesh.flags, BT_MESH_COMP_DIRTY) && (page == 0U || page == 1U)) {
		buf->size -= BT_MESH_MIC_SHORT;
		err = bt_mesh_comp_read(&sdu, page);
		if (err) {
			LOG_ERR("Unable to get stored composition data");
			return err;
		}

		buf->size += BT_MESH_MIC_SHORT;
		return 0;
	}

	if (page == 0 || page == 128) {
		bt_mesh_comp_data_get_page_0(buf, offset);
	} else if (page == 1 || page == 129) {
		bt_mesh_comp_data_get_page_1(buf);
	} else {
		return -EINVAL;
	}
}

...

err = comp_data_get_page(&sdu, page, 0);
if (err) {
	...
	return err;
}
```

Andrewpini replied:
Yes, this could work. My main issue when starting implementing this part was that it was a bit cluttery and a lot of redundant checks 👍

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1892,7 +1892,16 @@ static int comp_set(const char *name, size_t len_rd, settings_read_cb read_cb,
-BT_MESH_SETTINGS_DEFINE(comp, "cmp", comp_set);
+BT_MESH_SETTINGS_DEFINE(comp, "cmp", comp_p0_set);
+
+static int comp_p1_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+		    void *cb_arg)
+{
+	/* Identical behavior as for page 0 */
+	comp_p0_set(name, len_rd, read_cb, cb_arg);
+	return 0;
+}
+BT_MESH_SETTINGS_DEFINE(comp1, "cmp1", comp_p1_set);
```
If the behavior is identical, the set handler is probably not needed?
```suggestion
BT_MESH_SETTINGS_DEFINE(comp1, "cmp1", comp_p0_set);
```

Andrewpini replied:
I belive I had some issues with this:

I tried using the same CB for both, I.E. `BT_MESH_SETTINGS_DEFINE(comp1, "cmp1", comp_p0_set);` -> /zephyr/subsys/bluetooth/mesh/settings.h:30:20: error: redefinition of ‘pre_comp_p0_set’.

I also tried `BT_MESH_SETTINGS_DEFINE(comp1, "cmp1", NULL);` -> zephyr/subsys/bluetooth/mesh/access.c:1904:40: error: called object is not a function or function pointer.

So I was not able to find any other way to do this, any advice?

Andrewpini replied:
```
static int comp_set(const char *name, size_t len_rd, settings_read_cb read_cb,
		    void *cb_arg)
{
	if (!atomic_test_bit(bt_mesh.flags, BT_MESH_INIT)) {
		return 0;
	}

	/* Only need to know that the entry exists. Will load the contents on
	 * demand.
	 */
	if (len_rd > 0) {
		atomic_set_bit(bt_mesh.flags, BT_MESH_COMP_DIRTY);
	}

	return 0;
}

#ifdef CONFIG_BT_SETTINGS
SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_comp, "bt/mesh/cmp", NULL, comp_set, NULL, NULL);
SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_comp1, "bt/mesh/cmp1", NULL, comp_set, NULL, NULL);

#else
/* Declaring non static settings handler helps avoid unnecessary ifdefs
 * as well as unused function warning. Since the declared handler structure is
 * unused, linker will discard it.
 */
const struct settings_handler settings_handler_bt_mesh_comp = {
	.h_set = comp_set,
};

#endif
```

This is one alterantive that is possible. `BT_MESH_SETTINGS_DEFINE` is not designed to handle reuse of functions, so alternatively we need to create a separate macro in `settings.h` to handle this particular case.

Andrewpini replied:
To be honset im not sure if this bloat is worth it to avoid the redundant `comp1` set cb

PavelVPV replied:
What if we store CDPs like this: `bt/mesh/cmp/0`, `bt/mesh/cmp/1`, ...? Then the old code will work as is, but will be just called as many times as number of pages that were stored.

Andrewpini replied:
> What if we store CDPs like this: `bt/mesh/cmp/0`, `bt/mesh/cmp/1`, ...? Then the old code will work as is, but will be just called as many times as number of pages that were stored.

Will this be backwards compatible?

PavelVPV replied:
It is part of mesh-1.1 features, it shouldn't have been used before.

Andrewpini replied:
I mean, Alex mentioned earlier in this PR that changing how cmp0 is stored might cause issues if the new scheme is part of a DFU update. I just wanted to mention it since he won't be able to participate in the rest of this PR. But if you think that this is not an issue im fine with doing it like this 👍

PavelVPV replied:
Replied in that conversation

Andrewpini commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2111,13 +2136,20 @@ int bt_mesh_comp_change_prepare(void)
+		LOG_ERR("Failed to clear composition data P1: %d", err_p1);
+	}
+
+	if (!err_p0 && !err_p1) {
```
Hmm, good point. I tested it in shell, and it seems like `settings_delete`  returns 0 when there is no such entry.

PavelVPV commented:
@Andrewpini, maybe lets use the patch from this comment in this PR? https://github.com/Andrewpini/zephyr/pull/2#discussion_r1308423528

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2113,11 +2139,12 @@ static void comp_data_clear(void)
+	for (int i = 0; i < ARRAY_SIZE(comp_data_pages); i++) {
+		err = settings_delete(comp_data_pages[i].path);
+		if (err) {
+			LOG_ERR("Failed to clear composition data %d: %d", comp_data_pages[i].page,
```
```suggestion
			LOG_ERR("Failed to clear CDP%d: %d", comp_data_pages[i].page,
```

PavelVPV commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -75,27 +79,18 @@ static int dev_comp_data_get(struct bt_mesh_model *model,
+	}
+
+	if (err) {
+		LOG_ERR("Failed to get composition page %d, err:%d", page, err);
```
```suggestion
		LOG_ERR("Failed to get CDP%d, err:%d", page, err);
```

PavelVPV reviewed (commented):
Sorry, few small comments before approving.
